AMDGPU/GlobalISel: Disable LCSSA pass

petar-avramovic opened the pull request:
Disable LCSSA pass in preparation for implementing temporal divergence
lowering in amdgpu divergence lowering. Breaks all cases where sgpr or
i1 values are used outside of the cycle with divergent exit.
Regenerate regression tests for amdgpu divergence lowering with LCSSA
disabled.
Update IntrinsicLaneMaskAnalyzer to stop tracking lcssa phis that are
lane masks.

petar-avramovic commented:
* **#124299** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124299?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#124298** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124298?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#124297** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124297?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/124297?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1366,7 +1366,8 @@ bool GCNPassConfig::addPreISel() {
   addPass(createAMDGPURewriteUndefForPHILegacyPass());
 
-  addPass(createLCSSAPass());
+  if (!getCGPassBuilderOption().EnableGlobalISelOption)
```
You can't just disable a pass for only globalisel. This will break the dag if there is a fallback

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalizeHelper.cpp:
```diff
@@ -653,7 +678,8 @@ void RegBankLegalizeHelper::applyMappingPHI(MachineInstr &MI) {
   // Divergent G_PHIs have vgpr dst but inputs can be sgpr or vgpr.
-  if (Ty == LLT::scalar(32) || Ty == LLT::pointer(4, 64)) {
+  if (Ty == LLT::scalar(32) || Ty == LLT::pointer(1, 64) ||
+      Ty == LLT::pointer(4, 64)) {
```
Why does this need to maintain a list of types

rovka commented:
Can you please put a minimal example in the PR description where LCSSA breaks things for the temporal divergence lowering, and explain why divergence lowering can't handle it?

arsenm commented:
> Can you please put a minimal example in the PR description where LCSSA breaks things for the temporal divergence lowering, and explain why divergence lowering can't handle it?

It's not acceptable for the lowering to not handle it. Not-LCSSA is not a guaranteed property

rovka commented:
> > Can you please put a minimal example in the PR description where LCSSA breaks things for the temporal divergence lowering, and explain why divergence lowering can't handle it?
> 
> It's not acceptable for the lowering to not handle it. Not-LCSSA is not a guaranteed property

Ok, that was my impression too, but I wanted to see if I was missing smth :)

petar-avramovic commented:
>  Can you please put a minimal example in the PR description where LCSSA breaks things for the temporal divergence lowering, and explain why divergence lowering can't handle it?
 
 > It's not acceptable for the lowering to not handle it. Not-LCSSA is not a guaranteed property

GlobalISel temporal divergence lowering(next 2 patches) + new regbankselect(and regbank legalize) can handle both LCSSA  and Not-LCSSA. This was discussed some time ago that globalisel path should disable LCSSA since it is not required for correctness.
SDAG, the fix passes after isel that move-to-VALU, silently require LCSSA to deal with temporal divergence.

arsenm commented:
> GlobalISel temporal divergence lowering(next 2 patches) + new regbankselect(and regbank legalize) can handle both LCSSA and Not-LCSSA. This was discussed some time ago that globalisel path should disable LCSSA since it is not required for correctness. SDAG, the fix passes after isel that move-to-VALU, silently require LCSSA to deal with temporal divergence.

But that still breaks the fallback. This should be an optional step, and I'd prefer to do it after. It can also only do this in the case where the fallback is not enabled

petar-avramovic commented:
LCSSA disabled only for global-isel without fallback (this is the default) and with new-reg-bank-select.

arsenm reviewed (commented):
This seems to have merged in another patch?

petar-avramovic commented on llvm/lib/Target/AMDGPU/AMDGPURegBankLegalize.cpp:
```diff
@@ -312,6 +312,12 @@ bool AMDGPURegBankLegalize::runOnMachineFunction(MachineFunction &MF) {
+      RBLHelper.applyMappingTrivial(*MI);
+      continue;
+    }
+
```
moved to https://github.com/llvm/llvm-project/pull/128702

petar-avramovic committed:
AMDGPU/GlobalISel: Disable LCSSA pass

Disable LCSSA pass in preparation for implementing temporal divergence
lowering in amdgpu divergence lowering. Breaks all cases where sgpr or
i1 values are used outside of the cycle with divergent exit.
Regenerate regression tests for amdgpu divergence lowering with LCSSA
disabled.
Update IntrinsicLaneMaskAnalyzer to stop tracking lcssa phis that are
lane masks.

petar-avramovic commented:
### Merge activity

* **Mar 12, 6:09 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/124297).
